### DISCUSSION TITLE
Scripting a rebase in multiple repositories

### DISCUSSION THREAD

**Discussion by jcf** [magit/magit#4321] [Support] [answered]
Hello fellow Magit users!

I have a problem with scripting Git and Magit in Emacs. I would like to be able to update a set of repositories from a single keybinding, and have written the following function to do the heavy lifting for me:

``` emacs-lisp
;;;###autoload
(defun +box/update ()
  (interactive)
  (dolist (repo '("~/.config/box/"
                  "~/.config/doom/"
                  "~/.config/zsh/prezto/"
                  "~/.password-store/"
                  "~/code/words/"))
    (let ((default-directory repo))
      (message "Updating %s..." default-directory)
      (magit-pull-from-upstream (list "--autostash")))))
```

When I run this from a virtual buffer or from a file outside of a Git repository I get a lot of errors.

```
Updating ~/.config/box/...
Running git pull --autostash files refs/heads/master
Updating ~/.config/doom/...
Running git pull --autostash files refs/heads/master
Updating ~/.config/zsh/prezto/...
Running git pull --autostash files refs/heads/master
Updating ~/.password-store/...
Running git pull --autostash origin refs/heads/master
Updating ~/code/words/...
Running git pull --autostash github refs/heads/master
Git<1> finished
error in process sentinel: magit--not-inside-repository-error: Not inside Git repository
error in process sentinel: Not inside Git repository
Git<2> finished
error in process sentinel: magit--not-inside-repository-error: Not inside Git repository
error in process sentinel: Not inside Git repository
Git<3> finished
error in process sentinel: magit--not-inside-repository-error: Not inside Git repository
error in process sentinel: Not inside Git repository
Git finished
error in process sentinel: magit--not-inside-repository-error: Not inside Git repository
error in process sentinel: Not inside Git repository
Git<4> finished
error in process sentinel: magit--not-inside-repository-error: Not inside Git repository
error in process sentinel: Not inside Git repository
```

When I run this from a buffer visiting a file inside a Git repository it appears to do the trick.

```
Updating ~/.config/box/...
Running git pull --autostash files refs/heads/master
Updating ~/.config/doom/...
Running git pull --autostash files refs/heads/master
Updating ~/.config/zsh/prezto/...
Running git pull --autostash files refs/heads/master
Updating ~/.password-store/...
Running git pull --autostash origin refs/heads/master
Updating ~/code/words/...
Running git pull --autostash github refs/heads/master
Git finished
Git<1> finished
Git<2> finished
Git<3> finished
Git<4> finished
```

Given I'm setting `default-directory` (which Magit relies on) I'm suprised to see this sometimes fail. Why would the buffer I'm in have any bearing on the call to `magit-pull-from-upstream` with a lexically-bound `default-directory` in place?

The top of my Emacs Lisp file contains…

```
;;; private/box/autoload.el -*- lexical-binding: t; -*-
```

…in case that's relevant. I'm not sure it is. :slightly_smiling_face: 

I'll try adding some additional logging to Magit's internals now to see if `default-directory` is wrong.

Thank you in advance for any and all suggestions! This has been bugging me for ages.

**Comment by tarsius** [comment, marked as answer]
This command runs `git` asynchronously and the let-binding is not in effect anymore by the time it is done. At that time the status buffer of the "current repository" is refreshed. You can suppress this by let binding `inhibit-magit-refresh`. The value of this binding is explicitly stored in the process object and taken into account when the process finishes.

**Reply by jcf** [reply]
YES! Thank you so much, @tarsius! That did it. :boom: 

``` emacs-lisp
;;;###autoload
(defun +box/update ()
  (interactive)
  (dolist (repo '("~/.config/box/"
                  "~/.config/doom/"
                  "~/.config/zsh/prezto/"
                  "~/.password-store/"
                  "~/code/words/"))
    (let ((default-directory repo)
          (inhibit-magit-refresh t))
      (message "Updating %s..." default-directory)
      (magit-pull-from-upstream (list "--autostash")))))
```

With `inhibit-magit-refresh` in place I can see all the `git` processes firing up and I get no error output in Emacs. You're a star! Thank you for your help. :bow:

**Reply by tarsius** [reply]
You are welcome!

By the way, I took this opportunity to rename `inhibit-magit-refresh` to `magit-inhibit-refresh`. The old name will remain available for a few months at least but you might still want to adjust this in a few hours when this change hits melpa.
